File: go/sigil/cache_diagnostics_test.go
package sigil

import (
	"context"
	"testing"
)

func TestSetCacheDiagnostics_NilRecorder(t *testing.T) {
	SetCacheDiagnostics(nil, "system_changed")
}

func TestSetCacheDiagnostics_EmptyReason(t *testing.T) {
	client, _, _ := newTestClient(t, Config{})
	_, rec := client.StartGeneration(context.Background(), GenerationStart{
		Model: ModelRef{Provider: "anthropic", Name: "claude-3-5-sonnet-latest"},
	})
	SetCacheDiagnostics(rec, "   ")
	rec.SetResult(Generation{Output: []Message{AssistantTextMessage("ok")}}, nil)
	rec.End()
	if err := rec.Err(); err != nil {
		t.Fatalf("End: %v", err)
	}
	if rec.lastGeneration.Metadata[CacheDiagnosticsMissReasonKey] != nil {
		t.Fatalf("expected no miss_reason in metadata, got %v", rec.lastGeneration.Metadata[CacheDiagnosticsMissReasonKey])
	}
}

func TestSetCacheDiagnostics_StampsMetadata(t *testing.T) {
	client, _, _ := newTestClient(t, Config{})
	_, rec := client.StartGeneration(context.Background(), GenerationStart{
		Model: ModelRef{Provider: "anthropic", Name: "claude-3-5-sonnet-latest"},
	})
	SetCacheDiagnostics(rec, "tools_changed",
		WithMissedInputTokens(42),
		WithPreviousMessageID("msg_prev"),
	)
	rec.SetResult(Generation{Output: []Message{AssistantTextMessage("ok")}}, nil)
	rec.End()
	if err := rec.Err(); err != nil {
		t.Fatalf("End: %v", err)
	}
	md := rec.lastGeneration.Metadata
	if got, _ := md[CacheDiagnosticsMissReasonKey].(string); got != "tools_changed" {
		t.Fatalf("miss_reason: want tools_changed, got %v", md[CacheDiagnosticsMissReasonKey])
	}
	if got, _ := md[CacheDiagnosticsMissedInputTokensKey].(string); got != "42" {
		t.Fatalf("missed_input_tokens: want 42, got %v", md[CacheDiagnosticsMissedInputTokensKey])
	}
	if got, _ := md[CacheDiagnosticsPreviousMessageIDKey].(string); got != "msg_prev" {
		t.Fatalf("previous_message_id: want msg_prev, got %v", md[CacheDiagnosticsPreviousMessageIDKey])
	}
}

func TestSetCacheDiagnostics_OverridesPreviousCall(t *testing.T) {
	client, _, _ := newTestClient(t, Config{})
	_, rec := client.StartGeneration(context.Background(), GenerationStart{
		Model: ModelRef{Provider: "anthropic", Name: "claude-3-5-sonnet-latest"},
	})
	SetCacheDiagnostics(rec, "system_changed", WithMissedInputTokens(1))
	SetCacheDiagnostics(rec, "model_changed", WithMissedInputTokens(99))
	rec.SetResult(Generation{Output: []Message{AssistantTextMessage("ok")}}, nil)
	rec.End()
	md := rec.lastGeneration.Metadata
	if got, _ := md[CacheDiagnosticsMissReasonKey].(string); got != "model_changed" {
		t.Fatalf("miss_reason: want model_changed, got %v", md[CacheDiagnosticsMissReasonKey])
	}
	if got, _ := md[CacheDiagnosticsMissedInputTokensKey].(string); got != "99" {
		t.Fatalf("missed tokens: want 99, got %v", md[CacheDiagnosticsMissedInputTokensKey])
	}
}

File: js/test/cache-diagnostics.test.mjs
import assert from 'node:assert/strict';
import { trace } from '@opentelemetry/api';
import test from 'node:test';
import {
  CACHE_DIAGNOSTICS_MISSED_INPUT_TOKENS_KEY,
  CACHE_DIAGNOSTICS_MISS_REASON_KEY,
  CACHE_DIAGNOSTICS_PREVIOUS_MESSAGE_ID_KEY,
  defaultConfig,
  setCacheDiagnostics,
  SigilClient,
} from '../.test-dist/index.js';

class MockGenerationExporter {
  requests = [];

  async exportGenerations(request) {
    this.requests.push(structuredClone(request));
    return {
      results: request.generations.map((generation) => ({
        generationId: generation.id,
        accepted: true,
      })),
    };
  }

  async shutdown() {}
}

test('setCacheDiagnostics stamps metadata on generation', async () => {
  const exporter = new MockGenerationExporter();
  const defaults = defaultConfig();
  const client = new SigilClient({
    tracer: trace.getTracer('sigil-sdk-js-test'),
    generationExport: {
      ...defaults.generationExport,
      batchSize: 1,
      flushIntervalMs: 60_000,
    },
    generationExporter: exporter,
  });
  try {
    const rec = client.startGeneration({
      conversationId: 'conv-cd',
      model: { provider: 'anthropic', name: 'claude-3-5-sonnet-latest' },
    });
    setCacheDiagnostics(rec, 'messages_changed', {
      missedInputTokens: 500,
      previousMessageId: 'msg_prev',
    });
    rec.setResult({ output: [{ role: 'assistant', content: 'ok' }] });
    rec.end();
    assert.equal(rec.getError(), undefined);
    await client.flush();
    const snap = client.debugSnapshot();
    assert.equal(snap.generations.length, 1);
    const md = snap.generations[0].metadata ?? {};
    assert.equal(md[CACHE_DIAGNOSTICS_MISS_REASON_KEY], 'messages_changed');
    assert.equal(md[CACHE_DIAGNOSTICS_MISSED_INPUT_TOKENS_KEY], '500');
    assert.equal(md[CACHE_DIAGNOSTICS_PREVIOUS_MESSAGE_ID_KEY], 'msg_prev');
  } finally {
    await client.shutdown();
  }
});

test('setCacheDiagnostics ignores empty reason', async () => {
  const exporter = new MockGenerationExporter();
  const defaults = defaultConfig();
  const client = new SigilClient({
    tracer: trace.getTracer('sigil-sdk-js-test'),
    generationExport: {
      ...defaults.generationExport,
      batchSize: 1,
      flushIntervalMs: 60_000,
    },
    generationExporter: exporter,
  });
  try {
    const rec = client.startGeneration({
      model: { provider: 'anthropic', name: 'claude-3-5-sonnet-latest' },
    });
    setCacheDiagnostics(rec, '   ');
    rec.setResult({ output: [{ role: 'assistant', content: 'ok' }] });
    rec.end();
    await client.flush();
    const md = client.debugSnapshot().generations[0].metadata ?? {};
    assert.equal(md[CACHE_DIAGNOSTICS_MISS_REASON_KEY], undefined);
  } finally {
    await client.shutdown();
  }
});
